### ISSUE TITLE
incorrect quantiles() output

### ISSUE THREAD

**Issue by belm0** [carsonfarmer/streamhist#7]
quantiles output can be far out of the histogram input range:
```python
>>> h = streamhist.StreamHist()
>>> h.update((5, 7, 70))
>>> h.quantiles(.5, .9, .95)
[-24.5, 70.0, 70.0]
```

**Comment by belm0** [comment]
more bizarre: quantiles not monotonically increasing:
```
h.quantiles(.25, .5, .9, .95)
[4.5, -24.5, 70.0, 70.0]
```

**Comment by belm0** [comment]
... because `compute_breaks()` count is wrong (negative values appearing in series):
```python
>>> h.compute_breaks()[0][:5]
[  1.15, -0.14,  0.02,  0.02,  0.02]
```
... because `sum()` is wrong (not monotonically increasing):
```python
>>> h.compute_breaks()[1][:5]
[5.0, 6.3, 7.6, 8.9, 10.2]
>>> [h.sum(x) for x in h.compute_breaks()[1][:5]]
[  0.50,  1.15,  1.01,  1.03,  1.05]

**Comment by belm0** [comment]
The streamhist `sum()` implementation doesn't seem to match the algorithm in the paper.  Notably `bin_sums()` is averaging neighboring bins:
```
def bin_sums(array, less):
    return [(a.count + b.count)/2 for a, b in _izip(array[:-1], array[1:]) if  b.value <= less]
```
while Sum Procedure (Algorithm 3) in the paper shows to simply sum all bins prior to the query's bin, and then add half of query's bin:
```
for all j < i do
    s = s+mj
end for
s = s+mi/2
```

Running the paper's example in Appendix A against StreamHist confirms the bug, with sum(15) output .5 less than the correct value (2.78 vs. 3.28), due to the misapplied averaging in `bin_sums()`.
```python
>>> h = StreamHist(maxbins=5)
>>> for x in (23,19,10,16,36,2,9): h.update(x)
>>> h2 = StreamHist(maxbins=5)
>>> for x in (32,30,45): h2.update(x)
>>> h3 = h + h2
>>> h3.bins
[(2, 1), (9.5, 2), (19.33, 3), (32.66, 3), (45, 1)
>>> h3.sum(15)
2.7750
```

**Comment by belm0** [comment]
Noting that streamhist appears to be a direct port of https://github.com/bigmlcom/histogram, adopting the latter's bugs and deviation from the Ben-Haim paper.

Even with `sum()` corrected, `quantiles()` output is still incorrect-- I misspoke that `quantiles()` depends on `compute_breaks()`.  `quantiles() uses `bin_sums()` (which I left as-is using the faulty averaging), and overall the algorithm appears suspect.

**Comment by carsonfarmer** [comment]
I am unable to devote much time to this package these days. I am able to evaluate and merge PRs (thanks for those!), so if it would help your application, please feel free to submit a PR with the simplified/trivial quantiles implementation.
Related, if #3 is of interest to you, please let me know. I can probably find some time in the next week or two to push this out to pip.

**Comment by belm0** [comment]
> I am unable to devote much time to this package these days. I am able to evaluate and merge PRs (thanks for those!), so if it would help your application, please feel free to submit a PR with the simplified/trivial quantiles implementation.  Related, if #3 is of interest to you, please let me know.

Thank you.  Still thinking about how to proceed.  I've spent a lot of time with an implementation from scratch, trying to optimize update performance and produce accurate quantile results.  Summary of my learnings:
  * in plain Python with no dependencies, it's possible to have 5x faster update than streamhist.  It mainly comes from maintaining a cost array (diff between neighboring points) rather than recompute.
  * (there's no reason to have the sortedcollections dependency.  sortedcollections wouldn't really help until max_bins is >> 1,000 (10,000 is its internal list length), and by then computing the bins to merge is the overwhelming bottleneck anyway.  sortedcollections uses plain lists with insort, so we can do the same directly.)
  * (given the speed that's possible, any "freeze bins" feature is questionable.  Users will only be caught off guard if their distribution shifts over time and they get wildly inaccurate output.)
  * using numba I was able to achieve an additional 5x update speed (total 20x faster than streamhist).  Having a fast "find bins to merge" opened up a further optimization "just mutate existing bin if new bin would be subsequently merged anyway".  It could be offered as an optional implementation only when user has numba/numpy installed.
  * it's possible to implement a quantile function with output identical to numpy.quantile() up until max_bins is reached.  Then it gracefully transitions to approximation.  Key insight here was to treat bins of count one as "exact" rather than half before / half after.

Perhaps it's easier to build up what I have so far as a new package/API, especially since streamhist doesn't exist yet on pypi.   @carsonfarmer what would you like to see become of streamhist?

**Comment by cy-zheng** [comment]
Same issue. Wrong quantiles() result.

**Comment by belm0** [comment]
I wrote:
> it's possible to implement a quantile function with output identical to `numpy.quantile()` up until max_bins is reached. Then it gracefully transitions to approximation. Key insight here was to treat bins of count one as "exact" rather than half before / half after.

I've opened #11 contributing my improved quantile implementation.

**Comment by carsonfarmer** [comment]
Fixed with #11

**Comment by carsonfarmer** [comment]
@belm0 I see I never actually addressed an old question of yours. I think I probably should just add this to pipy so others can use it more easily. But realize this comment is coming many months too late. I have some time to publish now, but if you're sorted out an alternatively, let me know and I won't bother.

**Comment by belm0** [comment]
@carsonfarmer thank you.  I was using streamhist in a project, but have since removed the dependency.  I don't have an opinion on putting streamhist in pypi.

In my perf-timer library, I ended up making a local implementation of the Ben-Haim paper.  It depends only on stdlib, has 130 lines of code, and is about 5x faster than streamhist.

  https://github.com/belm0/perf-timer/blob/af9563f177fe4fa06270e9722d18b3790b90dc5e/src/perf_timer/_histogram.py

Recently I helped the distogram author with "exact quantile output when histogram is under capacity" (as we did with streamhist) and a performance improvement.
